### PR TITLE
WIP: Sanitize url and baseurl values on loading config

### PR DESCRIPTION
**Pull request by ashmaroli**
Benefits Jekyll's URLFilters &mdash; Jekyll doesn't have to sanitize the values on each call to the filter(s).

### Summary:
  - `config["url"]` and `config["baseurl"]` will be *reset* to `""` if the set value is `nil`, or an instance other than that of `String`, or if the value contains `".."`
  - `config["url"]` will have all *trailing* slash(es) removed
  - `config["baseurl"]` will have all *trailing* slash(es) and any extra slashes

### CODE REVIEW

**Comment by pathawks** [comment]
>`config["url"]` and `config["baseurl"]` will be reset to "" if the set value is nil, or an instance other than that of `String`, or if the value contains `".."`

I am wondering if it would be better for us to just error out of the build if `config["url"]` is an `Array` or something goofy like that. _Or_ maybe we should just `to_s` everything, in case the `baseurl` is `1234` instead of `"1234"`.

I'm not sure.

File: lib/jekyll/configuration.rb [hunk 1]
```diff
@@ -99,7 +99,7 @@ class << self
       # problems and backwards-compatibility.
       def from(user_config)
         Utils.deep_merge_hashes(DEFAULTS, Configuration[user_config].stringify_keys)
-          .add_default_collections
+          .add_default_collections.sanitize_url!.sanitize_baseurl!
```
Comment (pathawks) [inline lib/jekyll/configuration.rb]: Is it more readable if each of these is on a separate line?

File: lib/jekyll/configuration.rb [hunk 2]
```diff
@@ -285,6 +285,26 @@ def renamed_key(old, new, config, _ = nil)
       end
     end
 
+    def sanitize_url!
```
Comment (pathawks) [inline lib/jekyll/configuration.rb]: Should these be `private`? Why might a plugin want access to them?

File: lib/jekyll/configuration.rb (hunk 1 above)
Comment (ashmaroli) [reply lib/jekyll/configuration.rb]: Yes definitely. But, Rubocop complains that the class has exceeded the number of lines allowed..

**Comment by pathawks** [comment]
https://github.com/jekyll/jekyll/blob/925fb6da45cbfeefa2accb7fa3faf99803c48a2d/test/test_filters.rb#L570-L577

This test thinks we should not attempt to use `baseurl` if it is improperly set. I think, rather than silently ignoring a wrong `baseurl`, we should either warn the user that something is broken, or just fail to build.

https://github.com/jekyll/jekyll/blob/925fb6da45cbfeefa2accb7fa3faf99803c48a2d/test/test_filters.rb#L597-L601

This test thinks we should always run `to_s` on `baseurl`.

https://github.com/jekyll/jekyll/blob/925fb6da45cbfeefa2accb7fa3faf99803c48a2d/test/test_commands_serve.rb#L261-L270

This test expects `site.url` to always end in a trailing slash.

**Comment by ashmaroli** [comment]
> This test expects site.url to always end in a trailing slash.

You're wrong there.. :smiley:
The test in **`test_command_serve.rb`** tests to see if Jekyll overrides the given value (with `localhost:4000`) in `env("production")`. Yeah, the test doesn't make it obvious though..

**Comment by pathawks** [comment]
> The test in **`test_command_serve.rb`** tests to see if Jekyll overrides the given value (with `localhost:4000`) in `env("production")`.

That is the true purpose of the test, yes. A (perhaps unintentional) second function of the test is that it expects `site.url` to always have a trailing slash, and complains when it does not (as is currently the case). I think we can just remove the trailing slash from ` assert_equal "https://jekyllrb.com/"` and call it good; the test would still be making sure it isn't set to `localhost` or anything else.

**Comment by ashmaroli** [comment]
> it expects `site.url` to always have a trailing slash

#6947 reports that its not to good behavior to expect.. removing the trailing `/` in the test..

**Comment by ashmaroli** [comment]
@pathawks leaving out changes to URLFilters in this PR.. I think we'll need to decide how to handle situations where the `config` has been altered erroneously by a plugin..

File: test/test_commands_serve.rb [hunk 3]
```diff
@@ -262,10 +262,10 @@ def serve(opts)
         should "not update the site url" do
           expect(Jekyll).to receive(:env).and_return("production")
           expect(Jekyll::Commands::Serve).to receive(:start_up_webrick)
-          @merc.execute(:serve, { "watch" => false, "url" => "https://jekyllrb.com/" })
+          @merc.execute(:serve, { "watch" => false, "url" => "https://jekyllrb.com" })
```
Comment (pathawks) [inline test/test_commands_serve.rb]: This will probably break many sites in weird ways. We need to document this really well.

**Pull request closed without merging**